### ISSUE TITLE
`<variant>`: `visit<R>()` uses wrong `static_assert`

### ISSUE THREAD

**Issue by hewillk** [microsoft/STL#2970]
https://github.com/microsoft/STL/blob/b7e0365b5e2f9de738e8fbea42b6cc6880efe531/stl/inc/variant#L1650-L1653

`_Variant_all_visit_results_implicitly_convertible` requires `_Convertible_from_all` which require `is_convertible` which doesn't know about guaranteed elision.

```cpp
#include <variant>

struct A {
  A() = default;
  A(const A&) = delete;
};

struct B {
  A operator()();
};

int main() {
  std::visit<A>(
    [](auto f) { return f(); }, std::variant<B>{});
}
```
https://godbolt.org/z/EcMfM5ffz

**Comment by frederick-vs-ja** [comment]
Similar to LLVM-55346. I think we should use `_Invoke_convertible` in `_Convertible_from_all` instead of `is_convertible`.

https://github.com/microsoft/STL/blob/b7e0365b5e2f9de738e8fbea42b6cc6880efe531/stl/inc/type_traits#L1617-L1622

**Comment by hewillk** [comment]
> Similar to [LLVM-55346](https://llvm.org/PR55346). I think we should use `_Invoke_convertible` in `_Convertible_from_all` instead of `is_convertible`.
> 
Yes, whichs comes from [SO](https://stackoverflow.com/questions/72166085/why-does-stdis-invocable-r-reject-functions-returning-non-moveable-types/72166327#72166327).

**Comment by frederick-vs-ja** [comment]
I failed to completely fix this due to DevCom-10024578. Theoretically the fix will work once the fixed compiler version is released.

**Comment by CaseyCarter** [comment]
> I failed to completely fix this due to [DevCom-10024578](https://developercommunity.visualstudio.com/t/10024578). Theoretically the fix will work once the fixed compiler version is released.

The internal bug database thinks this bug is fixed in 17.3p3. Is that not the case?

**Comment by frederick-vs-ja** [comment]
> The internal bug database thinks this bug is fixed in 17.3p3. Is that not the case?

The bug of `__is_constructible` has been fixed. But the bug revealed by [this example](https://developercommunity.visualstudio.com/t/10024578#T-N10026550) hasn't got fixed, and `__is_convertible` still seems broken.
